src/track/beatmap.cpp

```diff
@@ -537,30 +554,125 @@ void BeatMap::translate(double dNumSamples) {
+    if (!isValid() || m_beats.isEmpty()) {
+        return;
+    }
+
+    switch (scale) {
+    case DOUBLE:
+        if (getBpm() * 2 > getMaxBpm()) {
+            return;
+        }
+        // introduce a new beat into every gap
+        scaleDouble();
+        break;
+    case HALVE:
+        // remove every second beat
+        scaleHalve();
+        break;
+    case TWOTHIRDS:
+        // introduce a new beat into every gap
+        scaleDouble();
+        // remove every second and third beat
+        scaleThird();
+        break;
+    case THREEFOURTHS:
+        // introduce two beats into every gap
+        scaleTriple();
+        // remove every second third and forth beat
+        scaleFourth();
+        break;
+    default:
+        DEBUG_ASSERT(!"scale value invalid");
         return;
     }
+    onBeatlistChanged();
+    locker.unlock();
+    emit(updated());
+}
+
+void BeatMap::scaleDouble() {
+    Beat prevBeat = m_beats.first();
+    BeatList::iterator it = m_beats.begin() + 1;
+    for (; it != m_beats.end(); ++it) {
+        // Need to not accrue fractional frames.
+        int distance = it->frame_position() - prevBeat.frame_position();
+        Beat beat;
+        beat.set_frame_position(prevBeat.frame_position() + distance / 2);
+        it = m_beats.insert(it, beat);
+        prevBeat = (++it)[0];
+    }
+}
 
-    // Scale the distance between every beat by 1/dScalePercentage to scale the
-    // BPM by dScalePercentage.
-    const double kScaleBeatDistance = 1.0 / dScalePercentage;
+void BeatMap::scaleTriple() {
     Beat prevBeat = m_beats.first();
     BeatList::iterator it = m_beats.begin() + 1;
     for (; it != m_beats.end(); ++it) {
         // Need to not accrue fractional frames.
-        double newFrame = floor(
-                (1 - kScaleBeatDistance) * prevBeat.frame_position() +
-                kScaleBeatDistance * it->frame_position());
-        it->set_frame_position(newFrame);
+        int distance = it->frame_position() - prevBeat.frame_position();
+        Beat beat;
+        beat.set_frame_position(prevBeat.frame_position() + distance / 3);
+        it = m_beats.insert(it, beat);
+        ++it;
+        beat.set_frame_position(prevBeat.frame_position() + distance * 2 / 3);
+        it = m_beats.insert(it, beat);
+        prevBeat = (++it)[0];
+    }
+}
+
+void BeatMap::scaleHalve() {
+    Beat prevBeat = m_beats.first();
+    BeatList::iterator it = m_beats.begin() + 1;
+    for (; it != m_beats.end(); ++it) {
+        it = m_beats.erase(it);
+        if (it == m_beats.end()) {
+            break;
+        }
+    }
+}
+
+void BeatMap::scaleThird() {
+    Beat prevBeat = m_beats.first();
+    BeatList::iterator it = m_beats.begin() + 1;
+    for (; it != m_beats.end(); ++it) {
+        it = m_beats.erase(it);
+        if (it == m_beats.end()) {
+            break;
+        }
+        it = m_beats.erase(it);
+        if (it == m_beats.end()) {
```

Comment:
I am afraid we will skip end() in this case and then  it == m_beats.end() is never true. 
We may fix it with it >= m_beats.end(); 
It can works here, but can be implementation dependent.